File: v2/lib/domain/stock.ts
// Logika stok kok — port dari server.js. Fungsi murni: hitung delta & validasi,
// biar repo layer yang nulis ke DB.

import type { Kok, KokType } from "./types";

export function parseStock(value: unknown, fallback = 0): number {
  const n = Number(value);
  if (!Number.isFinite(n)) return fallback;
  return Math.max(0, Math.round(n));
}

export function countKoksByType(koks: Kok[] | undefined): Map<string, number> {
  const map = new Map<string, number>();
  for (const k of koks ?? []) {
    const id = k?.typeId;
    if (!id) continue;
    map.set(id, (map.get(id) ?? 0) + 1);
  }
  return map;
}

/**
 * Delta stok per typeId untuk transisi oldKoks → newKoks.
 * Positif = stok nambah (kok dilepas), negatif = stok berkurang (kok dipakai).
 * Konsisten dgn applyKoksStockDiff lama: delta = oldCount − newCount.
 */
export function stockDeltas(
  oldKoks: Kok[] | undefined,
  newKoks: Kok[] | undefined,
): Map<string, number> {
  const oldMap = countKoksByType(oldKoks);
  const newMap = countKoksByType(newKoks);
  const ids = new Set([...oldMap.keys(), ...newMap.keys()]);
  const out = new Map<string, number>();
  for (const id of ids) {
    const delta = (oldMap.get(id) ?? 0) - (newMap.get(id) ?? 0);
    if (delta !== 0) out.set(id, delta);
  }
  return out;
}

/** Cek stok cukup buat selisih oldKoks→newKoks. Balik pesan error atau null. */
export function stockDiffError(
  kokTypes: KokType[],
  oldKoks: Kok[] | undefined,
  newKoks: Kok[] | undefined,
): string | null {
  const oldMap = countKoksByType(oldKoks);
  const newMap = countKoksByType(newKoks);
  for (const id of newMap.keys()) {
    const need = (newMap.get(id) ?? 0) - (oldMap.get(id) ?? 0);
    if (need <= 0) continue;
    const type = kokTypes.find((t) => t.id === id);
    if (!type) continue;
    const avail = Math.max(0, Number(type.stock) || 0);
    if (need > avail) {
      return `Stok ${type.name} tidak cukup (butuh ${need}, sisa ${avail})`;
    }
  }
  return null;
}

File: v3/server/internal/domain/stock.go
// Port dari v2/lib/domain/stock.ts — aritmetika stok kok. Fungsi murni,
// repo layer (F2) yang menulis hasilnya ke DB.
package domain

import "fmt"

type KokType struct {
	ID           string
	Name         string
	PricePerSlop int64
	Stock        int
	Active       bool
}

// CountKoksByType menghitung berapa kok per typeId. Kok tanpa TypeID diabaikan.
func CountKoksByType(koks []Kok) map[string]int {
	out := map[string]int{}
	for _, k := range koks {
		if k.TypeID == "" {
			continue
		}
		out[k.TypeID]++
	}
	return out
}

// StockDeltas menghitung delta stok per typeId untuk transisi oldKoks →
// newKoks. Positif = stok nambah (kok dilepas), negatif = stok berkurang
// (kok dipakai).
func StockDeltas(oldKoks, newKoks []Kok) map[string]int {
	oldMap := CountKoksByType(oldKoks)
	newMap := CountKoksByType(newKoks)
	out := map[string]int{}
	seen := map[string]bool{}
	for id := range oldMap {
		seen[id] = true
	}
	for id := range newMap {
		seen[id] = true
	}
	for id := range seen {
		delta := oldMap[id] - newMap[id]
		if delta != 0 {
			out[id] = delta
		}
	}
	return out
}

// StockDiffError mengecek stok cukup buat selisih oldKoks→newKoks. Balik
// pesan error atau string kosong kalau cukup.
func StockDiffError(kokTypes []KokType, oldKoks, newKoks []Kok) string {
	oldMap := CountKoksByType(oldKoks)
	newMap := CountKoksByType(newKoks)
	byID := map[string]KokType{}
	for _, t := range kokTypes {
		byID[t.ID] = t
	}
	for id, newCount := range newMap {
		need := newCount - oldMap[id]
		if need <= 0 {
			continue
		}
		typ, ok := byID[id]
		if !ok {
			continue
		}
		avail := typ.Stock
		if avail < 0 {
			avail = 0
		}
		if need > avail {
			return fmt.Sprintf("Stok %s tidak cukup (butuh %d, sisa %d)", typ.Name, need, avail)
		}
	}
	return ""
}

// SlopsFromStock — jumlah slop untuk stok pcs (ceil, min 1 kalau stock > 0).
// 1 slop = 12 kok.
func SlopsFromStock(stock int) int {
	if stock <= 0 {
		return 0
	}
	slops := (stock + 11) / 12
	if slops < 1 {
		return 1
	}
	return slops
}

// ExpenseFromInitialStock — kas keluar dari stok awal (pcs) + harga/slop.
func ExpenseFromInitialStock(stock int, pricePerSlop int64) int64 {
	if pricePerSlop <= 0 {
		return 0
	}
	return int64(SlopsFromStock(stock)) * pricePerSlop
}
